File: xjakt00-adventura/PrikazDej.java
/*******************************************************************************
 * Pomocí příkazu 'dej' může hráč předávat osobám určitě předměty.
 *
 * @author    Tomáš Jakubčin (xjakt00)
 * @version   pro školní rok 2010/11
 */
public class PrikazDej implements IPrikaz
{
    private static final String NAZEV = "dej";
    private Batoh batoh;
    private HerniPlan plan;
    private Hra hra;

    /***************************************************************************
     *  Konstruktor třídy.
     *  
     *  @param batoh odkaz na hráčův batoh.
     *  @param plan  odkaz na herní plán.
     *  @param hra   odkaz na hru.
     */ 
    public PrikazDej(Batoh batoh, HerniPlan plan, Hra hra) {
        this.batoh = batoh;
        this.plan = plan;
        this.hra = hra;
    }

    /**
     *  Zkontroluje parametry (musí být dva - osoba a věc, kterou chce hráč předat)
     *  a předá věc.
     */ 
    public String proved (String... parametry) {
        Prostor kdeJsme = plan.getAktualniProstor();
        if (parametry.length == 0) {
            return "Nevim, komu mam co dat.";
        }
        String jmenoKomuDat = parametry [0];
        String nazevCoDat;
        try {
            nazevCoDat = parametry [1];
        }
        catch (java.lang.ArrayIndexOutOfBoundsException e) {
            return "Nevim, co mam dat - musis zadat dva parametry";
        }
        Osoba osoba = kdeJsme.vyberOsobu(jmenoKomuDat);
        Vec vecicka = batoh.vyberVec(nazevCoDat);
        if (osoba == null) {
            return "Tato osoba v prostoru neni.";
        }
        if (vecicka == null) {
            return "Tuto vec nemas v batohu";
        }
        if (osoba.maVec(nazevCoDat) == true) {
            return "Osoba uz tuto vec ma a dalsi nejspis nepotrebuje.";
        }
        if (plan.getCas()-1 > 0) {
            plan.setCas(3);
            batoh.setPocetVeci(-1);
            batoh.setVaha(-vecicka.getVaha());
            vecicka.setVBatohu(-1);
            osoba.setVec(vecicka);
            if (vecicka.getVBatohu() <= 0) {
                batoh.removeVec(vecicka);
            }
            return "Dal jsi vec '"+nazevCoDat+"' osobe '"+osoba.getJmeno()+"'.";
        }
        else {
            plan.konecHry();
            return "";
        }
    }

    /**
     *  Metoda vrací název příkazu (slovo které používá hráč pro jeho vyvolání)
     *  
     *  @ return nazev prikazu
     */
    public String getNazev() {
        return NAZEV;
    }

}

File: src/Adventura/PrikazDej.cpp
#include "PrikazDej.h"

#include "Hra.h"

PrikazDej::PrikazDej(const std::string & nazev, Hra * hra, HerniPlan * herni_plan, Batoh * batoh)
	: IPrikaz(nazev), odkaz_na_hru(hra), odkaz_na_plan(herni_plan), odkaz_na_batoh(batoh)
{
}

PrikazDej::~PrikazDej()
{
}

void PrikazDej::proved(const std::vector<std::string> & parametry)
{
	if (parametry.size() == 0)
	{
		std::cout << "Nevim komu a co mam dat.\n";
	}
	else if (parametry.size() == 1)
	{
		std::cout << "Nevim co mam dat.\n";
	}
	else if (parametry.size() == 2)
	{
		auto aktualni_prostor = odkaz_na_plan->get_aktualni_prostor();
		if (aktualni_prostor->je_osoba_v_prostoru(parametry[0]) == false)
		{
			std::cout << "Nelze predat veci osobe, ktera neni v prostoru.\n";
			return;
		}
		
		bool je_v_batohu = odkaz_na_batoh->mam_vec(parametry[1]);
		if (je_v_batohu == false)
		{
			std::cout << "Nelze predat vec, kterou nemas v batohu.\n";
			return;
		}

		auto osoba = aktualni_prostor->get_osoba(parametry[0]);
		osoba->pridej_vec(parametry[1], 1);
		odkaz_na_batoh->odeber_vec(parametry[1]);

		std::cout << "Dal jsi vec " << parametry[1] << " osobe " << parametry[0] << '\n';
	}
	else
	{
		std::cout << "Nelze dat vice veci najednou.\n";
	}
}

std::string PrikazDej::nazev(void) const
{
	return IPrikaz::jmeno_prikazu;
}

std::string PrikazDej::pouziti(void)const
{
	return "dej <komu> <co>";
}